llmkernel/kernel.py

```diff
@@ -32,107 +61,281 @@ class PythonLLMKernel(IPythonKernel):
         "file_extension": ".txt",
     }
 
+    internal_executions: set[str]
+    subkernel_execution_tracking: dict[str, str]
+
+    def __init__(self, server):
+        self.internal_executions = set()
+        self.subkernel_execution_tracking = {}
+        self.subkernel_id = None
+        super().__init__(server)
+        self.new_kernel()
+        self.setup_instance()
+
+    def setup_instance(self):
+        self.setup_llm()
+        self.setup_intercepts()
 
-    def setup_instance(self, *args, **kwargs):
-        self.toolset = None
-        self.agent = None
+    def setup_llm(self):
         # Init LLM agent
+        if getattr(self, 'context', None) is not None:
+            self.agent.clear_all_context()
         self.context = None
-        self.msg_types.append("context_setup_request")
-        self.msg_types.append("llm_request")
-        self.msg_types.append("download_dataset_request")
-        self.msg_types.append("save_dataset_request")
-        self.msg_types.append("save_amr_request")
-        self.msg_types.append("load_dataset")
-        self.msg_types.append("load_mira_model")
-        return super().setup_instance(*args, **kwargs)
-
-
-    def set_context(self, context, context_info):
-        self.toolset = None
+
+    def setup_intercepts(self):
+        self.server.intercept_message("shell", "context_setup_request", self.context_setup_request)
+        self.server.intercept_message("shell", "llm_request", self.llm_request)
+        self.server.intercept_message("shell", "execute_request", self.track_execute_request)
+        self.server.intercept_message("iopub", "execute_input", self.update_execute_input_response)
+        self.server.intercept_message("shell", "execute_reply", self.post_execute)
+
+    def connect_to_last(self):
+        # We don't want to automatically connect to the last kernel
+        return
+    
+    def new_kernel(self):
+        # Shutdown any existing subkernel (if it exists) before spinnup up a new kernel
+        self.shutdown_subkernel()
```

Comment:
Why not use the `cleanup` function here?

Reply:
`cleanup` is exterior to the kernel class itself and runs as a final step after the ioloop is ended. We might be able to clean this up a bit, but it would probably be the other way around, calling shutdown_subkernel() directly instead of cleanup.